File: common/src/main/java/de/nicidienase/chaosflix/common/entities/recording/ConferencesWrapper.java
package de.nicidienase.chaosflix.common.entities.recording;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Created by felix on 17.03.17.
 */

public class ConferencesWrapper implements Parcelable{
	private static final String TAG = ConferencesWrapper.class.getSimpleName();
	private static final String CONGRESS = "congress";
	private static final String EVENTS = "events";
	private static final String CONFERENCES = "conferences";
	private static final String DEFAULT_CONFERENCE_GROUP = "other conferences";
	private static final String EVENT_GROUP = "Events";
	private static final int MIN_NUM_CONS = 1;

	private List<Conference> conferences;
	private Map<String, List<Conference>> conferenceMap = null;

	protected ConferencesWrapper(Parcel in) {
		conferences = in.createTypedArrayList(Conference.CREATOR);
	}

	public static final Creator<ConferencesWrapper> CREATOR = new Creator<ConferencesWrapper>() {
		@Override
		public ConferencesWrapper createFromParcel(Parcel in) {
			return new ConferencesWrapper(in);
		}

		@Override
		public ConferencesWrapper[] newArray(int size) {
			return new ConferencesWrapper[size];
		}
	};

	public static String getStringForTag(String tag) {
		switch (tag) {
			case "congress":
				return "Congress";
			case "sendezentrum":
				return "Sendezentrum";
			case "camp":
				return "Camp";
			case "broadcast/chaosradio":
				return "Chaosradio";
			case "eh":
				return "Easterhegg";
			case "gpn":
				return "GPN";
			case "froscon":
				return "FrOSCon";
			case "mrmcd":
				return "MRMCD";
			case "sigint":
				return "SIGINT";
			case "datenspuren":
				return "Datenspuren";
			case "fiffkon":
				return "FifFKon";
			case "blinkenlights":
				return "Blinkenlights";
			case "chaoscologne":
				return "1c2 Chaos Cologne";
			case "cryptocon":
				return "CryptoCon";
			case "other conferences":
				return "Other Conferences";
			case "denog":
				return "DENOG";
			case "vcfb":
				return "Vintage Computing Festival Berlin";
			case "hackover":
				return "Hackover";
			case "netzpolitik":
				return "Das ist Netzpolitik!";
			default:
				return tag;
		}
	}

	public static List<String> getOrderedConferencesList() {
		return Arrays.asList("congress", "sendezentrum", "camp",
				"gpn", "mrmcd", "broadcast/chaosradio",
				"eh", "froscon", "sigint",
				"datenspuren", "fiffkon", "cryptocon");
	}


	public List<Conference> getConferences() {
		return conferences;
	}

	public Map<String, List<Conference>> getConferencesBySeries() {
		if (conferenceMap == null) {
			conferenceMap = new HashMap<>();
			for (Conference conference : conferences) {
				String[] split = conference.getSlug().split("/");
				List<Conference> list;
				switch (split[0]) {
					case CONGRESS:
						if (split[1].endsWith("sendezentrum")) {
							getListForTag("sendezentrum").add(conference);
						} else {
							getListForTag(CONGRESS).add(conference);
						}
						break;
					case CONFERENCES:
						switch (split.length) {
							case 2:
								if (split[1].startsWith("camp")) {
									getListForTag("camp").add(conference);
								} else if (split[1].startsWith("sigint")) {
									getListForTag("sigint").add(conference);
								} else if (split[1].startsWith("eh")) {
									getListForTag("eh").add(conference);
								} else {
									getListForTag(DEFAULT_CONFERENCE_GROUP).add(conference);
								}
								break;
							case 3:
								getListForTag(split[1]).add(conference);
								break;
							default:
								getListForTag(DEFAULT_CONFERENCE_GROUP).add(conference);
								break;
						}
						break;
					case EVENTS:
						getListForTag(EVENT_GROUP).add(conference);
						break;
					default:
						getListForTag(conference.getSlug()).add(conference);
				}
			}
		}
		List<Conference> other = conferenceMap.get(DEFAULT_CONFERENCE_GROUP);
		Set<String> keySet = conferenceMap.keySet();
		List<String> removeList = new ArrayList<>();
		for (String tag : keySet) {
			if (!tag.equals(DEFAULT_CONFERENCE_GROUP)) {
				List<Conference> list = conferenceMap.get(tag);
				Collections.sort(list);
				Collections.reverse(list);
				if (list.size() <= MIN_NUM_CONS) {
					Log.d(TAG, "To few conferences: " + tag);
					other.addAll(list);
					removeList.add(tag);
				}
			}
		}
		for (String key : removeList) {
			conferenceMap.remove(key);
		}
		return conferenceMap;
	}

	private List<Conference> getListForTag(String s) {
		if (conferenceMap != null) {
			if (conferenceMap.keySet().contains(s)) {
				return conferenceMap.get(s);
			} else {
				List<Conference> list = new ArrayList<>();
				conferenceMap.put(s, list);
				return list;
			}
		}
		return null;
	}

	public void setConferences(List<Conference> conferences) {
		this.conferences = conferences;
	}

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeTypedList(conferences);
		dest.writeString(CONGRESS);
		dest.writeString(EVENTS);
		dest.writeString(CONFERENCES);
		dest.writeString(DEFAULT_CONFERENCE_GROUP);
		dest.writeString(EVENT_GROUP);
		dest.writeInt(MIN_NUM_CONS);
	}
}

File: common/src/main/java/de/nicidienase/chaosflix/common/entities/recording/ConferencesWrapper.kt
package de.nicidienase.chaosflix.common.entities.recording

import java.util.*
import kotlin.collections.HashMap


data class ConferencesWrapper(var conferences: List<Conference>) {
    private val CONGRESS = "congress"
    private val EVENTS = "events"
    private val CONFERENCES = "conferences"
    private val DEFAULT_CONFERENCE_GROUP = "other conferences"
    private val EVENT_GROUP = "Events"
    private val MIN_NUM_CONS = 1

    val conferencesMap: Map<String, List<Conference>>
        get() = generateConferencesMap()

    private fun generateConferencesMap(): HashMap<String, MutableList<Conference>> {
        val map = HashMap<String,MutableList<Conference>>()
        for (conference in conferences) {
            val split = conference.slug.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            when (split[0]) {
                CONGRESS -> if (split[1].endsWith("sendezentrum")) {
                    getListForTag(map,"sendezentrum").add(conference)

                } else {
                    getListForTag(map,CONGRESS).add(conference)
                }
                CONFERENCES -> when (split.size) {
                    2 -> if (split[1].startsWith("camp")) {
                        getListForTag(map,"camp").add(conference)
                    } else if (split[1].startsWith("sigint")) {
                        getListForTag(map,"sigint").add(conference)
                    } else if (split[1].startsWith("eh")) {
                        getListForTag(map,"eh").add(conference)
                    } else {
                        getListForTag(map,DEFAULT_CONFERENCE_GROUP).add(conference)
                    }
                    3 -> getListForTag(map,split[1]).add(conference)
                    else -> getListForTag(map,DEFAULT_CONFERENCE_GROUP).add(conference)
                }
                EVENTS -> getListForTag(map,EVENT_GROUP).add(conference)
                else -> getListForTag(map,conference.slug).add(conference)
            }
        }
        val other = map[DEFAULT_CONFERENCE_GROUP]
        val keySet = map.keys
        val removeList = ArrayList<String>()
        for (tag in keySet) {
            if (tag != DEFAULT_CONFERENCE_GROUP) {
                val list = map[tag]
                Collections.sort(list)
                Collections.reverse(list)
                if (list!!.size <= MIN_NUM_CONS) {
                    other?.addAll(list)
                    removeList.add(tag)
                }
            }
        }
        for (key in removeList) {
            map.remove(key)
        }
        return map
    }

    private fun getListForTag(map: MutableMap<String,MutableList<Conference>>, s: String): MutableList<Conference> {
        if (map.keys.contains(s)) {
            return map[s]!!
        } else {
            val list = ArrayList<Conference>()
            map.put(s, list)
            return list
        }
    }

    companion object {
        fun getStringForTag(tag: String): String {
            when (tag) {
                "congress" -> return "Congress"
                "sendezentrum" -> return "Sendezentrum"
                "camp" -> return "Camp"
                "broadcast/chaosradio" -> return "Chaosradio"
                "eh" -> return "Easterhegg"
                "gpn" -> return "GPN"
                "froscon" -> return "FrOSCon"
                "mrmcd" -> return "MRMCD"
                "sigint" -> return "SIGINT"
                "datenspuren" -> return "Datenspuren"
                "fiffkon" -> return "FifFKon"
                "blinkenlights" -> return "Blinkenlights"
                "chaoscologne" -> return "1c2 Chaos Cologne"
                "cryptocon" -> return "CryptoCon"
                "other conferences" -> return "Other Conferences"
                "denog" -> return "DENOG"
                "vcfb" -> return "Vintage Computing Festival Berlin"
                "hackover" -> return "Hackover"
                "netzpolitik" -> return "Das ist Netzpolitik!"
                else -> return tag
            }
        }

        val orderedConferencesList: List<String> = Arrays.asList(
                "congress", "sendezentrum", "camp",
                "gpn", "mrmcd", "broadcast/chaosradio",
                "eh", "froscon", "sigint",
                "datenspuren", "fiffkon", "cryptocon")
    }
}
